File: JavaScript solutions/1769. Minimum Number of Operations to Move All Balls to Each Box.js
//approach :1
/**
 * @param {string} boxes
 * @return {number[]}
 */
var minOperations = function (boxes) {
  let n = boxes.length;
  let pref = new Array(n).fill(0);
  let ones = boxes[0] === "1" ? 1 : 0;

  for (let i = 1; i < n; i++) {
    pref[i] = pref[i - 1] + ones;
    if (boxes[i] === "1") ones++;
  }

  ones = 0;
  let prev = 0;

  for (let i = n - 1; i >= 0; i--) {
    pref[i] += prev;
    if (boxes[i] === "1") ones++;
    prev += ones;
  }

  return pref;
};

// approach :2
function minOperations(boxes) {
  const len = boxes.length;
  const v = [];

  for (let i = 0; i < len; i++) {
    if (boxes[i] === "0") {
      v.push(0);
    } else {
      v.push(i + 1);
    }
  }

  const res = [];

  for (let i = 0; i < len; i++) {
    let ans = 0;

    for (let j = 0; j < len; j++) {
      if (v[j] !== 0) {
        ans += Math.abs(i - v[j] + 1);
      }
    }

    res.push(ans);
  }

  return res;
}

File: C++ solutions/1769. Minimum Number of Operations to Move All Balls to Each Box.cpp
#include<bits/stdc++.h>
using namespace std;

// solution : 1
class Solution {
public:
    vector<int> minOperations(string boxes) {
        int n = boxes.size();
        vector<int> pref(n,0);
        int ones = boxes[0] == '1';
        for(int i = 1; i < n; ++i){
            pref[i] = pref[i-1] + ones;
            if(boxes[i] == '1') ones++;
        }
        ones = 0;
        int prev = 0;
        for(int i = n-1; i >= 0; --i){
            pref[i] = pref[i] + prev;
            if(boxes[i] == '1'){
                ones++;
            }
            prev += ones;
        }
        return pref;
    }
};


// solution :2
class Solution {
public:
    vector<int> minOperations(string boxes) {
        int len=boxes.size();
        vector<int>v;
        for(int i=0;i<len;i++){
            if(boxes[i]=='0') v.push_back(0);
            else v.push_back(i+1);
        }
        vector<int>res;
        for(int i=0;i<len;i++){
            int ans=0;
            for(int j=0;j<len;j++){
                if(v[j]!=0) ans+=abs((i-v[j]+1));
            }
            res.push_back(ans);
        }
        return res;
    }
};